content/posts/looker-pdt-to-dbt-part-one.md

```diff
@@ -0,0 +1,184 @@
+
+The workflow described above is common place in dbt. Ultimately, it creates a separation between the preparation of data and the LookML that instructs Looker how to expose your final modelled output. 
+
+## Testing your models
+
+Another feature that dbt brings to the table is model testing, similar to unit tests present in most software development languages. 
+
+Looker's data governance capabilities are fantastic. They ensure that everyone is looking at the same data - the same metrics. However, it doesn't ensure that the data feeding those metrics is correct. If someone makes a mistake in the SQL, everyone is still looking at the same metric, that metric just isn't accurate anymore. 
+
+Imagine the following situation. You use Salesforce as your CRM and you have a LookML view called `sf_account`:
+
+{{< highlight sql >}}
+
+select 
+	a.id,
+	a.name
+from salesforce.sf_account a
+
+{{< /highlight >}}
+
+Everything is working correctly. You're completely confident in the data generated in your reports.
+
+Then, you decide you want to add the stage name from the related Salesforce opportunities to the model. Absent-mindedly, you forget that Accounts -> Opportunities is a one-to-many relationship. You make the following change and all your data from the `sf_account` view is suddently fanned out, distorting metrics in all your reports. 
+
+{{< highlight sql >}}
+
+select 
+	a.id,
+	a.name
+from salesforce.sf_account a
+left join salesforce.sf_opportunity o
+	on o.accountid = a.id
+
+{{< /highlight >}}
+
+The ID column was supposed to be your unique primary key and now it isn't. There is currently no automated mechanism in Looker to test the output of that query. No way of knowing that a record for ID `176` is now there 5 times, one for each opportunity generated for that account.
+
+In dbt, you can define constraints on the model, giving you the ability to identify instances like these in your development workflow.
+
+{{< highlight yaml >}}
+
+- name: sf_account
+  columns:
+    - name: id
+      tests:
+        - not_null
+        - unique
+
+{{< /highlight >}}
+
+When you run `dbt test`, you'd realise your mistake and fix it before your Head of Sales starts shouting at you because all the numbers changed. Though testing of that sort hasn't always been common-place in analytics, it's incredibly important so that everyone can have confidence in the reporting they consume.
+
+## PDT table names
+
+David from your data science team wants to build a fancy model, predicting how many flying mattresses you'll sell next week. To do so, he wants to use the `customer_status` dimension defined in your `customers` view in Looker. He doesn't want to have re-define it in his work. He understandably thinks you should all be using the same definition. 
```

Comment:
Very tiny thing, but can you change this to a woman's name since DS is a male-dominated industry? It's a pattern I've noticed recently in other blogs and really appreciate.

Reply:
```suggestion
 David from your data science team wants to build a fancy model, predicting how many flying mattresses you'll sell next week. To do so, he wants to use the `customer_status` dimension defined in your `customers` view in Looker. He doesn't want to have redefine it in his work. He understandably thinks you should all be using the same definition. 
```